ethcore/src/miner/stratum.rs

```diff
@@ -0,0 +1,233 @@
+			StratumServiceError::Dispatch(format!("{}", e))
+		)?;
+
+		trace!(
+			target: "stratum",
+			"submit_work: Decoded: nonce={}, pow_hash={}, mix_hash={}",
+			payload.nonce,
+			payload.pow_hash,
+			payload.mix_hash,
+		);
+
+		self.with_core_void(|client, miner| {
+			let seal = vec![encode(&payload.mix_hash).to_vec(), encode(&payload.nonce).to_vec()];
+			if let Err(e) = miner.submit_seal(&*client, payload.pow_hash, seal) {
+				warn!(target: "stratum", "submit_seal error: {:?}", e);
+			};
+		});
+
+		Ok(())
+	}
+}
+
+impl StratumJobDispatcher {
+	/// New stratum job dispatcher given the miner and client
+	fn new(miner: Weak<Miner>, client: Weak<Client>) -> StratumJobDispatcher {
+		StratumJobDispatcher {
+			seed_compute: Mutex::new(SeedHashCompute::new()),
+			client: client,
+			miner: miner,
+		}
+	}
+
+	/// Serializes payload for stratum service
+	fn payload(&self, pow_hash: H256, difficulty: U256, number: u64) -> String {
+		// TODO: move this to engine
+		let target = Ethash::difficulty_to_boundary(&difficulty);
+		let seed_hash = &self.seed_compute.lock().get_seedhash(number);
+		let seed_hash = H256::from_slice(&seed_hash[..]);
+		format!(
+			r#"["0x", "0x{}","0x{}","0x{}","0x{:x}"]"#,
+			pow_hash.hex(), seed_hash.hex(), target.hex(), number
+		)
+	}
+
+	fn with_core<F, R>(&self, f: F) -> Option<R> where F: Fn(Arc<Client>, Arc<Miner>) -> Option<R> {
+		self.client.upgrade().and_then(|client| self.miner.upgrade().and_then(|miner| (f)(client, miner)))
+	}
+
+	fn with_core_void<F>(&self, f: F) where F: Fn(Arc<Client>, Arc<Miner>) {
+		self.client.upgrade().map(|client| self.miner.upgrade().map(|miner| (f)(client, miner)));
+	}
+}
+
+/// Wrapper for dedicated stratum service
+pub struct Stratum {
+	dispatcher: Arc<StratumJobDispatcher>,
+	service: Arc<StratumService>,
+}
+
+#[derive(Debug)]
+/// Stratum error
+pub enum Error {
+	/// IPC sockets error
+	Service(StratumServiceError),
+	/// Invalid network address
+	Address(AddrParseError),
+}
+
+impl From<StratumServiceError> for Error {
+	fn from(service_err: StratumServiceError) -> Error { Error::Service(service_err) }
+}
+
+impl From<AddrParseError> for Error {
+	fn from(err: AddrParseError) -> Error { Error::Address(err) }
+}
+
+impl super::work_notify::NotifyWork for Stratum {
+	fn notify(&self, pow_hash: H256, difficulty: U256, number: u64) {
+		self.service.push_work_all(
+			self.dispatcher.payload(pow_hash, difficulty, number)
+		).unwrap_or_else(
+			|e| warn!(target: "stratum", "Error while pushing work: {:?}", e)
+		);
+	}
+}
+
+impl Stratum {
+
+	/// New stratum job dispatcher, given the miner, client and dedicated stratum service
+	pub fn start(options: &StratumOptions, miner: Weak<Miner>, client: Weak<Client>) -> Result<Stratum, Error> {
+		let dispatcher = Arc::new(StratumJobDispatcher::new(miner, client));
+
+		let stratum_svc = StratumService::start(
+			&SocketAddr::from_str(&format!("{}:{}", options.listen_addr, options.port))?,
```

Comment:
port and listen address are configured separately all over the configuration code
https://github.com/ethcore/parity/blob/master/parity/configuration.rs

but the SocketAddr instantiation should not be done via string, yes